docs-samples/V4/JS/manage-conversation-and-user-state/azure-table-storage.js

```diff
@@ -0,0 +1,83 @@
+ * 
+ * To run this bot:
+ * 1) install these npm packages:
+ * npm install --save restify
+ * npm install --save botbuilder@preview
+ * npm install --save botbuilder-azure@preview
+ * 
+ * 2) From VSCode, open the package.json file and
+ * Update the property "main" to the name of the sample bot you want to run. 
+ *    For example: "main": "manage-conversation-and-user-state/azure-table-storage.js" to run the this sample bot.
+ * 3) run the bot in debug mode. 
+ * 4) Load the emulator and point it to: http://localhost:3978/api/messages
+ * 5) Send the message "hi" to engage with the bot.
+ *
+ */ 
+
+const { BotFrameworkAdapter, ConversationState, UserState, BotStateSet } = require('botbuilder');
+const restify = require('restify');
+const { TableStorage } = require('botbuilder-azure');
+
+
+// Create server
+let server = restify.createServer();
+server.listen(process.env.port || process.env.PORT || 3978, function () {
+    console.log(`${server.name} listening to ${server.url}`);
+});
+
+// Create adapter (it's ok for MICROSOFT_APP_ID and MICROSOFT_APP_PASSWORD to be blank for now)  
+const adapter = new BotFrameworkAdapter({ 
+    appId: process.env.MICROSOFT_APP_ID, 
+    appPassword: process.env.MICROSOFT_APP_PASSWORD 
+});
+
+// Add conversation state middleware
+// The parameters are the connection string and table name.
+// "UseDevelopmentStorage=true" is the connection string to use if you are using the Azure Storage Emulator.
+// Replace it with your own connection string if you're not using the emulator
+var azureStorage = new TableStorage({ tableName: "TestAzureTable1", storageAccountOrConnectionString: "UseDevelopmentStorage=true"})
+
+// You can alternatively use your account name and table name
+// var azureStorage = new TableStorage({tableName: "TestAzureTable2", storageAccessKey: "V3ah0go4DLkMtQKUPC6EbgFeXnE6GeA+veCwDNFNcdE6rqSVE/EQO/kjfemJaitPwtAkmR9lMKLtcvgPhzuxZg==", storageAccountOrConnectionString: "storageaccount"});
+
+const conversationState = new ConversationState(azureStorage);
+const userState  = new UserState(azureStorage);
+adapter.use(conversationState);
+
+// Listen for incoming requests 
+server.post('/api/messages', (req, res) => {
+    // Route received request to adapter for processing
+    adapter.processActivity(req, res, async (context) => {
+        const isMessage = (context.activity.type === 'message');
+        const convo = conversationState.get(context);
+        const user = userState.get(context);
+
+        if (isMessage) {
+            if(!user.name && !convo.haveAskedNameFlag){
+                // Ask for the name.
+                await context.sendActivity("What is your name?")
+                // Set flag to show we've asked for the name. We save this out so the
+                // context object for the next turn of the conversation can check haveAskedNameFlag
+                convo.haveAskedNameFlag = true;
+            } else if(convo.haveAskedNameFlag){
```

Comment:
Setting individual flags like this is inefficient. You'll end up having a ton of flags floating around in our storage bag. 
Try using a "convo.prompt = 'haveAskedName'" or "convo.prompt = 'haveAskNumber'" 
To check if you're in the prompt: if(convo.prompt.match(/haveAskedName/ig)){
...do stufff
convo.prompt = undefined; // or an empty string ""
}

Reply:
Or, perhaps we should go ahead and use the "topicStates" object that I'm using in my primitive samples...that way we stay consistent when it comes time to using flags...:
// Define flags to manage the conversation flow and prompt states
var topicStates = {
    "topicTitle": undefined, // Current conversation topic in progress
    "prompt": undefined,     // Current prompt state in progress - indicate what question is being asked.
    "resetStates": function() {
        topicStates.topicTitle = undefined;
        topicStates.prompt = undefined;
    }
}